Fix country region select with autocomplete

louwie17 opened the pull request:
Fixes #5523 

Fixes some issues with the country / region dropdown field, in particular when using auto complete.
Issues that should be fixed:
- After using autocomplete, not being able to change the country/region anymore manually
- Having the chrome auto fill overlap the typeahead search results of the country/region field
- Region didn't support abbreviations (my autocomplete has NS for Nova Scotia, but the dropdown defaulted to New Brunswick)

### Screenshots

![country-region-autofill](https://user-images.githubusercontent.com/2240960/129060616-2f0e492e-ee74-400c-93fc-75264192acdc.gif)

### Detailed test instructions:

- Start a new WooCommerce store and go to **WooCommerce > Home** (make sure you have autofill enabled and some addresses configured)
- The onboarding flow should start, asking for the store address
- Do a autofill of the store address and make sure the Country/Region is correctly selected
- Now try each of these cases:
     - After using autocomplete, search for and select a country/region manually, it should change on click
     - Chrome should not show it's autofill ([as shown here](https://user-images.githubusercontent.com/90977/97976265-4f866200-1dc2-11eb-8546-e96f70cb9fad.png)) 
     - Clear the fields and make sure one of your auto complete addresses uses a abbreviated region (California -> CA) and see if it correctly selects the region with the country
- Feel free to try other things with the type ahead field and see if any odd behaviour happens and in other browsers like Safari or Firefox

<!--
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note using the following CLI command `npm run changelogger -- add` and commit changes. --->

louwie17 committed:
initial refactor

louwie17 committed:
Fix auto fill and write tests

louwie17 committed:
Removed autofill from country/region and added support for abbreviation regions

louwie17 commented on client/dashboard/components/settings/general/store-address.js:
```diff
@@ -200,6 +215,7 @@ export function StoreAddress( props ) {
 			<SelectControl
 				label={ __( 'Country / Region', 'woocommerce-admin' ) }
 				required
+				autoComplete="new-password" // disable autocomplete and autofill
```
This seems to be the only way to have the chrome autofill now show up. Chrome ignores the use of `autoComplete='off'` 😢

joshuatf replied:
Nice workaround!

louwie17 commented on client/dashboard/components/settings/general/store-address.js:
```diff
@@ -86,6 +86,18 @@ export function useGetCountryStateAutofill( options, countryState, setValue ) {
+			setAutofillCountry( newCountry );
+			setAutofillState( newState );
+		}
+	}, [ countryState ] );
```
Handle the `countryState` (user typing) state differently from the auto complete fields.
This makes it easier to know which is actually happening.

joshuatf replied:
I'm having some trouble parsing what this hook is for.  Could you elaborate a bit on this?

louwie17 replied:
This is only triggered when the `countryState` is updated, which means the user manually selected a country/region from the typeahead field (they didn't use autofill).
In this case, to keep consistent with the autofill, we update the autofill fields with the selected item from the dropdown, which is mostly what is happening in this hook.

louwie17 committed:
Add changelog

louwie17 committed:
Add changelog for the component package

joshuatf commented on client/dashboard/components/settings/general/store-address.js:
```diff
@@ -99,12 +111,15 @@ export function useGetCountryStateAutofill( options, countryState, setValue ) {
+				escapeRegExp( autofillState.replace( /\s/g, '' ) ) + '$', // Always match the end of string for region.
 				'i'
 			);
+			const isStateAbbreviation = autofillState.length < 3;
```
Do we need to consider country abbreviations?

louwie17 replied:
Probably not a bad idea, I only noticed the issue with state, but that was because I am using `Canada` and it's abbreviation matches nicely.
I will add that.

joshuatf reviewed (commented):
Love the workaround here, @louwie17!  Left a few questions in the code.

I also noticed that if the field is already filled, autocomplete doesn't work when selecting a new address.  This feels a bit odd to me, but is this expected?

louwie17 commented:
> I also noticed that if the field is already filled, autocomplete doesn't work when selecting a new address. This feels a bit odd to me, but is this expected?

@joshuatf from what I remember it was expected, when I tried to select a new address it would only change the field I was specifically on and the `onChange` handlers would not be triggered for any of the other fields (if they were filled out, my browser gave the option to clear them first and then select a new address which worked fine).
Was this the case for you? Meaning, when you selected a different address did the city & postal code fields change? or did they remain the same? (if they were previously filled).

louwie17 committed:
Fix clear form for autofill

louwie17 commented:
@joshuatf are you able to have another look at this? I added abbreviation support for countries and fixed the clear form on autofill.

One thing to note, the way my autofill works is: If you modified a field after using autofill, and then try to clear it with autofill, it will only clear the fields that hadn't been modified.

joshuatf commented on packages/components/src/select-control/index.js:
```diff
@@ -559,6 +563,7 @@ SelectControl.defaultProps = {
 	showClearButton: false,
 	hideBeforeSearch: false,
 	staticList: false,
+	autoComplete: 'off',
```
Should this default to `new-password` since `off` doesn't work with Chrome?  Or do you think consumers should be responsible for implementing this?

louwie17 replied:
I think consumers should still be responsible for this, as `new-password` isn't a natural way of turning it off. And from what I read, Chrome could decide to use `off` again if they so choose (`off` works fine on Safari and Firefox).

joshuatf reviewed (approved):
Testing well for my uses cases, thanks for nice solution, @louwie17.

Left one question about a default value, but a change might no be necessary and pre-approving ✅
